kimchi/src/tests/framework.rs

```diff
@@ -31,128 +31,122 @@ type BaseSponge = DefaultFqSponge<VestaParameters, SpongeParams>;
-                .unwrap();
-        println!("- time to create proof: {:?}s", start.elapsed().as_secs());
+    #[must_use]
+    pub(crate) fn public_inputs(mut self, public_inputs: Vec<Fp>) -> Self {
+        self.public_inputs = public_inputs;
+        self
+    }
 
-        // verify the proof
-        let start = Instant::now();
-        verify::<Affine, BaseSponge, ScalarSponge>(&group_map, &verifier_index, &proof).unwrap();
-        println!("- time to verify: {}ms", start.elapsed().as_millis());
+    #[must_use]
+    pub(crate) fn lookup_tables(mut self, lookup_tables: Vec<LookupTable<Fp>>) -> Self {
+        self.lookup_tables = lookup_tables;
+        self
     }
 
-    /// Create and verify a recursive proof
-    pub(crate) fn run_test_recursion(
-        gates: Vec<CircuitGate<Fp>>,
-        witness: [Vec<Fp>; COLUMNS],
-        public: &[Fp],
-    ) {
-        // create the index
-        let start = Instant::now();
-        let index = new_index_for_test(gates, public.len());
-        let verifier_index = index.verifier_index();
-        println!("- time to create index: {:?}s", start.elapsed().as_secs());
+    #[must_use]
+    pub(crate) fn runtime_tables_cfg(
+        mut self,
+        runtime_tables_cfg: Vec<RuntimeTableConfiguration>,
+    ) -> Self {
+        self.runtime_tables_cfg = Some(runtime_tables_cfg);
+        self
+    }
 
-        // verify the circuit satisfiability by the computed witness
-        index.cs.verify(&witness, public).unwrap();
-
-        // previous opening for recursion
-        let rng = &mut StdRng::from_seed([0u8; 32]);
-        let prev_challenges = {
-            let k = math::ceil_log2(index.srs.g.len());
-            let chals: Vec<_> = (0..k).map(|_| Fp::rand(rng)).collect();
-            let comm = {
-                let coeffs = b_poly_coefficients(&chals);
-                let b = DensePolynomial::from_coefficients_vec(coeffs);
-                index.srs.commit_non_hiding(&b, None)
-            };
-            (chals, comm)
-        };
+    #[must_use]
+    pub(crate) fn runtime_tables(mut self, runtime_tables: Vec<RuntimeTable<Fp>>) -> Self {
+        self.runtime_tables = runtime_tables;
+        self
+    }
 
-        // add the proof to the batch
-        let start = Instant::now();
-        let group_map = <Affine as CommitmentCurve>::Map::setup();
-        let proof = ProverProof::create_recursive::<BaseSponge, ScalarSponge>(
-            &group_map,
-            witness,
-            &[],
-            &index,
-            vec![prev_challenges],
-        )
-        .unwrap();
-        println!("- time to create proof: {:?}s", start.elapsed().as_secs());
+    #[must_use]
+    pub(crate) fn recursion(mut self, recursion: Vec<(Vec<Fp>, PolyComm<Affine>)>) -> Self {
+        self.recursion = recursion;
+        self
+    }
 
-        // verify the proof
+    /// creates the indexes
+    #[must_use]
+    pub(crate) fn setup(mut self) -> Self {
```

Comment:
It feels like this should return a different type, since you shouldn't continue 'setting it up' after calling this.